File: backend/app/models/slack.py
"""
Slack Integration Schemas
Pydantic models for Slack API data structures.
"""

from typing import List, Dict, Any, Optional
from datetime import datetime
import os
from pydantic import BaseModel, Field
from dotenv import load_dotenv

# Load environment variables from .env file
load_dotenv()

class SlackCredentialsSchema(BaseModel):
    """Schema for Slack API credentials."""
    bot_token: str = Field(default=os.getenv("SLACK_BOT_TOKEN", ""), description="Bot User OAuth Token from Slack")
    app_token: Optional[str] = Field(default=os.getenv("SLACK_APP_TOKEN", None), description="App-level token for advanced features")
    signing_secret: Optional[str] = Field(default=os.getenv("SLACK_SIGNING_SECRET", None), description="Signing secret for verifying requests")
    client_id: Optional[str] = Field(default=os.getenv("SLACK_CLIENT_ID", None), description="Client ID for OAuth")
    client_secret: Optional[str] = Field(default=os.getenv("SLACK_CLIENT_SECRET", None), description="Client Secret for OAuth")
    
    class Config:
        schema_extra = {
            "example": {
                "bot_token": "",
                "app_token": "",
                "signing_secret": "",
                "client_id": "",
                "client_secret": ""
            }
        }


class SlackUserSchema(BaseModel):
    """Schema for Slack user data."""
    id: str = Field(..., description="Slack user ID")
    name: str = Field(..., description="Slack username")
    real_name: Optional[str] = Field(None, description="User's real name")
    email: Optional[str] = Field(None, description="User's email address")
    title: Optional[str] = Field(None, description="User's job title")
    is_admin: bool = Field(False, description="Whether the user is a workspace admin")


class SlackMessageSchema(BaseModel):
    """Schema for Slack message data."""
    id: str = Field(..., description="Message timestamp ID")
    text: str = Field(..., description="Message text content")
    user: str = Field(..., description="User ID who sent the message")
    timestamp: datetime = Field(..., description="Message timestamp")
    reactions: List[Dict[str, Any]] = Field(default_factory=list, description="Message reactions")
    replies: List["SlackMessageSchema"] = Field(default_factory=list, description="Thread replies")
    thread_ts: Optional[str] = Field(None, description="Parent thread timestamp if in a thread")
    attachments: List[Dict[str, Any]] = Field(default_factory=list, description="Message attachments")


class SlackChannelSchema(BaseModel):
    """Schema for Slack channel data."""
    id: str = Field(..., description="Channel ID")
    name: str = Field(..., description="Channel name")
    is_private: bool = Field(False, description="Whether the channel is private")
    created: datetime = Field(..., description="Channel creation time")
    creator: Optional[str] = Field(None, description="User ID of channel creator")
    messages: List[SlackMessageSchema] = Field(default_factory=list, description="Channel messages")


class SlackCanvasSchema(BaseModel):
    """Schema for Slack canvas data."""
    id: str = Field(..., description="Canvas ID")
    title: str = Field(..., description="Canvas title")
    created: datetime = Field(..., description="Canvas creation time")
    creator: str = Field(..., description="User ID of canvas creator")
    content: List[Dict[str, Any]] = Field(default_factory=list, description="Canvas content blocks")
    channel_id: Optional[str] = Field(None, description="ID of associated channel")


class SlackWorkspaceSchema(BaseModel):
    """Schema for comprehensive Slack workspace data."""
    id: str = Field(..., description="Workspace ID")
    name: str = Field(..., description="Workspace name")
    domain: str = Field(..., description="Workspace domain")
    users: List[SlackUserSchema] = Field(default_factory=list, description="Workspace users")
    channels: List[SlackChannelSchema] = Field(default_factory=list, description="Workspace channels")
    canvases: List[SlackCanvasSchema] = Field(default_factory=list, description="Workspace canvases")
    extraction_time: datetime = Field(..., description="When the data was extracted")


class SlackAIProcessingRequestSchema(BaseModel):
    """Schema for requesting AI processing of Slack data."""
    workspace_id: str = Field(..., description="ID of the workspace to process")
    process_channels: List[str] = Field(default_factory=list, description="List of channel IDs to process (empty for all)")
    process_canvases: bool = Field(True, description="Whether to process canvas data")
    max_message_age_days: Optional[int] = Field(None, description="Maximum age of messages to process in days")
    instruction: Optional[str] = Field(None, description="Custom instructions for the AI processor")


class SlackAIProcessingResultSchema(BaseModel):
    """Schema for AI processing results of Slack data."""
    workspace_id: str = Field(..., description="ID of the processed workspace")
    summary: str = Field(..., description="Overall summary of the workspace")
    channel_summaries: Dict[str, str] = Field(default_factory=dict, description="Summaries for each channel")
    canvas_summaries: Dict[str, str] = Field(default_factory=dict, description="Summaries for each canvas")
    key_insights: List[str] = Field(default_factory=list, description="Key insights extracted")
    topics: List[Dict[str, Any]] = Field(default_factory=list, description="Topics identified across the workspace")
    processing_time: datetime = Field(..., description="When the processing was completed")


# Update forward references for nested models
SlackMessageSchema.update_forward_refs()

File: frontend/lib/api/slack.ts
/**
 * Slack API Client
 * Functions for interacting with the Slack integration backend API
 */

import axios from 'axios';

// Types
export interface SlackCredentials {
  bot_token: string;
  app_token?: string;
  signing_secret?: string;
  client_id?: string;
  client_secret?: string;
}

export interface SlackWorkspace {
  id: string;
  name: string;
  domain: string;
  users: SlackUser[];
  channels: SlackChannel[];
  canvases: SlackCanvas[];
  extraction_time: string;
}

export interface SlackUser {
  id: string;
  name: string;
  real_name?: string;
  email?: string;
  title?: string;
  is_admin: boolean;
}

export interface SlackChannel {
  id: string;
  name: string;
  is_private: boolean;
  created: string;
  creator?: string;
  messages: SlackMessage[];
}

export interface SlackMessage {
  id: string;
  text: string;
  user: string;
  timestamp: string;
  reactions: any[];
  replies: SlackMessage[];
  thread_ts?: string;
  attachments: any[];
}

export interface SlackCanvas {
  id: string;
  title: string;
  created: string;
  creator: string;
  content: any[];
  channel_id?: string;
}

export interface SlackAIProcessingRequest {
  workspace_id: string;
  process_channels?: string[];
  process_canvases?: boolean;
  max_message_age_days?: number;
  instruction?: string;
}

export interface SlackAIProcessingResult {
  workspace_id: string;
  summary: string;
  channel_summaries: Record<string, string>;
  canvas_summaries: Record<string, string>;
  key_insights: string[];
  topics: Array<{
    name: string;
    description: string;
    keywords: string[];
  }>;
  processing_time: string;
}

export interface ProcessingStatus {
  status: 'queued' | 'processing' | 'completed' | 'failed';
  progress?: number;
  error?: string;
}

// Connect to a Slack workspace using provided credentials
export async function connectSlackWorkspace(credentials: SlackCredentials) {
  const response = await axios.post('/slack/connect', credentials);
  return response.data;
}

// List all connected Slack workspaces
export async function listSlackWorkspaces() {
  const response = await axios.get('/slack/workspaces');
  return response.data;
}

// Extract data from a Slack workspace
export async function extractSlackData(workspaceId: string) {
  const response = await axios.post(`/slack/extract?workspace_id=${workspaceId}`);
  return response.data as SlackWorkspace;
}

// Process Slack data with AI
export async function processSlackData(request: SlackAIProcessingRequest) {
  const response = await axios.post('/slack/process', request);
  return response.data;
}

// Get processing status
export async function getProcessingStatus(taskId: string): Promise<ProcessingStatus> {
  const response = await axios.get(`/slack/process/status?task_id=${taskId}`);
  return response.data as ProcessingStatus;
}

// Get processing results
export async function getProcessingResults(taskId: string): Promise<SlackAIProcessingResult> {
  const response = await axios.get(`/slack/process/results?task_id=${taskId}`);
  return response.data as SlackAIProcessingResult;
}

// Poll processing status until complete
export const pollProcessingStatus = (
  taskId: string, 
  onStatusUpdate?: (status: ProcessingStatus) => void,
  intervalMs: number = 2000,
  maxAttempts: number = 30
): Promise<SlackAIProcessingResult> => {
  let attempts = 0;
  
  return new Promise((resolve, reject) => {
    const checkStatus = async () => {
      try {
        const status = await getProcessingStatus(taskId);
        
        if (onStatusUpdate) {
          onStatusUpdate(status);
        }
        
        if (status.status === 'completed') {
          const results = await getProcessingResults(taskId);
          resolve(results);
          return;
        }
        
        if (status.status === 'failed') {
          reject(new Error(status.error || 'Processing failed'));
          return;
        }
        
        attempts++;
        if (attempts >= maxAttempts) {
          reject(new Error('Processing timed out'));
          return;
        }
        
        setTimeout(checkStatus, intervalMs);
      } catch (err: any) {
        reject(new Error(`Error checking status: ${err.message}`));
      }
    };
    
    checkStatus();
  });
};
